File: scripts/analyze-duplicates.py
#!/usr/bin/env python3
"""
Analyze duplicate detection scan results.

Usage:
    python scripts/analyze-duplicates.py [report_path]

If no report path provided, uses the most recent scan report.
"""

import json
import sys
from collections import defaultdict
from pathlib import Path


def get_file_path(location):
    """Extract file path from location (handles both dict and string formats)."""
    if isinstance(location, dict):
        return location.get("file_path", "unknown")
    elif isinstance(location, str):
        return location.split(":")[0]
    return "unknown"


def analyze_report(report_path):
    """Analyze a duplicate detection report."""
    with open(report_path, "r") as f:
        data = json.load(f)

    # Group blocks by content_hash to find exact duplicates
    hash_to_blocks = defaultdict(list)
    for b in data.get("code_blocks", []):
        hash_to_blocks[b["content_hash"]].append(b)

    # Get groups with 2+ occurrences, sorted by count
    exact_dups = [(h, blocks) for h, blocks in hash_to_blocks.items() if len(blocks) >= 2]
    exact_dups.sort(key=lambda x: len(x[1]), reverse=True)

    print(f"Report: {report_path}")
    print(f"Total code blocks: {len(data.get('code_blocks', []))}")
    print(f"Exact duplicate patterns: {len(exact_dups)}")
    print()

    for i, (h, blocks) in enumerate(exact_dups[:15], 1):
        files = list(set(get_file_path(b["location"]) for b in blocks))
        code = blocks[0]["source_code"]
        short_files = [Path(f).name for f in files[:4]]

        print(f"=== Group {i}: {len(blocks)}x in {len(files)} files ===")
        print(f"Files: {', '.join(short_files)}")
        if len(files) > 4:
            print(f"       +{len(files)-4} more")
        print(f"Code:\n{code}\n")


def find_latest_report():
    """Find the most recent scan report."""
    output_dir = Path("output/reports")
    if not output_dir.exists():
        return None

    reports = list(output_dir.glob("scan-*.json"))
    if not reports:
        return None

    return max(reports, key=lambda p: p.stat().st_mtime)


if __name__ == "__main__":
    if len(sys.argv) > 1:
        report_path = sys.argv[1]
    else:
        report_path = find_latest_report()
        if not report_path:
            print("No scan reports found in output/reports/")
            sys.exit(1)

    analyze_report(report_path)

File: scripts/analysis/analyze-duplicates.ts
#!/usr/bin/env node --strip-types
/**
 * Analyze duplicate detection scan results.
 *
 * Usage:
 *   node --strip-types scripts/analysis/analyze-duplicates.ts [report_path]
 *
 * If no report path provided, uses the most recent scan report.
 */

import { readFileSync, readdirSync, statSync, existsSync } from 'node:fs';
import { basename, join } from 'node:path';

interface Location {
  file_path?: string;
}

interface CodeBlock {
  content_hash: string;
  location: Location | string;
  source_code: string;
}

interface ScanReport {
  code_blocks?: CodeBlock[];
}

const MAX_GROUPS_SHOWN = 15;
const MAX_FILES_INLINE = 4;
const MIN_DUPLICATE_COUNT = 2;

function getFilePath(location: Location | string): string {
  if (typeof location === 'object' && location !== null) {
    return location.file_path ?? 'unknown';
  }
  if (typeof location === 'string') {
    return location.split(':')[0];
  }
  return 'unknown';
}

function analyzeReport(reportPath: string): void {
  const raw = readFileSync(reportPath, 'utf-8');
  const data: ScanReport = JSON.parse(raw);
  const codeBlocks = data.code_blocks ?? [];

  const hashToBlocks = new Map<string, CodeBlock[]>();
  for (const block of codeBlocks) {
    const existing = hashToBlocks.get(block.content_hash);
    if (existing) {
      existing.push(block);
    } else {
      hashToBlocks.set(block.content_hash, [block]);
    }
  }

  const exactDups = [...hashToBlocks.entries()]
    .filter(([, blocks]) => blocks.length >= MIN_DUPLICATE_COUNT)
    .sort((a, b) => b[1].length - a[1].length);

  console.log(`Report: ${reportPath}`);
  console.log(`Total code blocks: ${codeBlocks.length}`);
  console.log(`Exact duplicate patterns: ${exactDups.length}`);
  console.log();

  for (let i = 0; i < Math.min(exactDups.length, MAX_GROUPS_SHOWN); i++) {
    const [, blocks] = exactDups[i];
    const files = [...new Set(blocks.map(b => getFilePath(b.location)))];
    const code = blocks[0].source_code;
    const shortFiles = files.slice(0, MAX_FILES_INLINE).map(f => basename(f));

    console.log(`=== Group ${i + 1}: ${blocks.length}x in ${files.length} files ===`);
    console.log(`Files: ${shortFiles.join(', ')}`);
    if (files.length > MAX_FILES_INLINE) {
      console.log(`       +${files.length - MAX_FILES_INLINE} more`);
    }
    console.log(`Code:\n${code}\n`);
  }
}

function findLatestReport(): string | null {
  const outputDir = 'output/reports';
  if (!existsSync(outputDir)) return null;

  const reports = readdirSync(outputDir)
    .filter(f => f.startsWith('scan-') && f.endsWith('.json'))
    .map(f => join(outputDir, f));

  if (reports.length === 0) return null;

  return reports.reduce((latest, current) =>
    statSync(current).mtimeMs > statSync(latest).mtimeMs ? current : latest
  );
}

const reportPath = process.argv[2] ?? findLatestReport();
if (!reportPath) {
  console.error('No scan reports found in output/reports/');
  process.exit(1);
}

analyzeReport(reportPath);
